Split CNode::cs_vSend: message processing and message sending

TheBlueMatt opened the pull request:
cs_vSend is used for two purposes - to lock the datastructures used
to queue messages to place on the wire and to only call
SendMessages once at a time per-node. I believe SendMessages used
to access some of the vSendMsg stuff, but it doesn't anymore, so
these locks do not need to be on the same mutex, and also make
deadlocking much more likely.

TheBlueMatt committed:
Split CNode::cs_vSend: message processing and message sending

cs_vSend is used for two purposes - to lock the datastructures used
to queue messages to place on the wire and to only call
SendMessages once at a time per-node. I believe SendMessages used
to access some of the vSendMsg stuff, but it doesn't anymore, so
these locks do not need to be on the same mutex, and also make
deadlocking much more likely.

TheBlueMatt commented:
Note that this commit was pulled out of #9419 as it stands alone and is by itself a big win.

theuni commented on src/net.cpp:
```diff
@@ -1887,7 +1883,7 @@ void CConnman::ThreadMessageHandler()
             // Send messages
             {
-                TRY_LOCK(pnode->cs_vSend, lockSend);
+                TRY_LOCK(pnode->cs_sendProcessing, lockSend);
```
No need for the TRY_LOCK here. Death to 'em!

TheBlueMatt replied:
Ok

TheBlueMatt committed:
Make the cs_sendProcessing a LOCK instead of a TRY_LOCK

Technically cs_sendProcessing is entirely useless now because it
is only ever taken on the one MessageHandler thread, but because
there may be multiple of those in the future, it is left in place

theuni commented:
utACK 11290734ca7261f732c9f43f152c69de3a42546c, I ran a nearly identical version for over a week with no issues.

Shameless beg/plea for 0.14. This was originally intended for #9441, but left out because #9419 (which has since been dropped for 0.14) already included it. I understand that we all have deep review queues atm though, not everything can make it.

TheBlueMatt commented:
Rebased after #9441 merge.

sipa commented:
utACK 376b3c2c6e329357e4793c1d1b90d1dc0f30fed0. Happy to see TRY_LOCKs go.

instagibbs commented:
lock pleb utACK 376b3c2c6e329357e4793c1d1b90d1dc0f30fed0

pouyapeyravi commented:
hi

morcos commented:
utACK 376b3c2 

similar to @instagibbs, this is relatively new code to me, but it seems like it makes sense

jtimon commented on src/net.cpp:
```diff
@@ -1272,12 +1270,10 @@ void CConnman::ThreadSocketHandler()
+                LOCK(pnode->cs_vSend);
+                size_t nBytes = SocketSendData(pnode);
+                if (nBytes) {
+                    RecordBytesSent(nBytes);
```
RecordBytesSent() needs to be called from inside the lock? If so, I assume the same would be needed in CConnman::PushMessage

theuni replied:
@jtimon No, RecordBytesSent does not require the lock. It has its own.

TheBlueMatt replied:
Just double checked, it looks like every variable accessed in RecordBytesSent is only done with cs_totalBytesSent (and even if this weren't true, this change wouldnt be a regression, we still call PushMessage from other functions without cs_vSend).

jtimon commented:
I was told by @TheBlueMatt and @theuni:
```
<BlueMatt> well #9535 is just a lock split, so just going through each variable that is accessed in one side and making sure its not accessed on the other is a pretty good (though admittedly not 100% sufficient) review
<BlueMatt> and one one of the sides in this case there are relatively few variables accessed, so its not so hard
<cfields> BlueMatt: yep, will do after dinner
<cfields> jtimon: yea, it's sane, just needs manual review of what's accessed under the locks before/after
<cfields> (note that you can treat it as non-recursive, so very simple)
```

This is what my "manual review  what's accessed under the locks before/after" contains at this point:
```
*** TODO #9535
****** Variables accessed while LOCK(pnode->cs_vSend) or TRY_LOCK(pnode->cs_vSend, lockSend);
pnode->vSendMsg
pnode->nSendOffset
pnode->hSocket
pnode->nLastSend
pnode->nSendBytes
pnode->nSendSize
pnode->fPauseSend
CConnman::nSendBufferMaxSize
pnode->mapSendBytesPerMsgCmd
pnode->cs_inventory
****** Variables accessed while LOCK(pnode->cs_sendProcessing)
GetNodeSignals().SendMessages(pnode, *this, flagInterruptMsgProc);
pto->nVersion
pto->fDisconnect
pto->GetSendVersion()
pto->fPingQueued
pto->nPingNonceSent
pto->nPingUsecStart
connman.PushMessage(pto, msgMaker.Make(NetMsgType::PING, nonce));
pto->nPingNonceSent
pto->GetId()
pto->fWhitelisted
pto->fAddnode
pto->fDisconnect
pto->addr
AdvertiseLocal(pto
pto->nNextLocalAddrSend
pto->nNextAddrSend
pto->vAddrToSend
pto->fClient
pto->fOneShot
```
L2685

...so far the method seems to be working, but would be nice to hear a confirmation that that's what I was told to confirm that I'm not making any wrong assumption. Only SendMessages() left and already listed functions/methods to review, but it's long, the tempation of doing a reduced version of #4646 always come back whenever asked to review SendMessages()...sorry.

At this point I doubt I will make my review before 0.14 is forked, but I'm fully motivated to finish it before or after it is merged.

theuni commented:
@jtimon looks good.

Re #4646, something similar is definitely the plan. See https://github.com/theuni/bitcoin/commit/1a6b10aea129ac363727c2d68fae809c2861e4da. That's definitely the next step, it was just too much of a refactor for 0.14.

As for the non-recursive part, I simply meant that it should now be quite easy to spot potential deadlocks since cs_vSend is now limited enough in scope that it can't recurse. In other words, the lock never leaves PushMessage/SocketSendData/RecordBytesSent (where RecordBytesSent isn't necessary and could be avoided).

jtimon commented:
Alright, going through all pnode->cs_vSend was easy, but after going though all SendMessages (I updated the list in my comment above), looking only at the pto parameter for the first pass (which is the most relevant for this pr) and not being exhaustive about all potential globals and not going recursively though connman.PushMessage() IsInitialBlockDownload(), Broadcast(nTimeBestReceived, &connman), pto->PushInventory(), pto->pfilter->IsRelevantAndUpdate()...the only "conflict" so far is cs_inventory, but it's just a finer smaller lock, so seems fine as well.
I'm very sorry about not being able to give a full utACK for this one, I feel my ACK is worth way more than the simple concept ACK I was planning and I thank @TheBlueMatt and @theuni for pushing me to review further when I warned I'm not so familiar with the networking code and its concurrency.
Slow review ACK 

re #4646 and https://github.com/theuni/bitcoin/commit/1a6b10aea129ac363727c2d68fae809c2861e4da I was thinking of something way simpler like #5153 but complete (I turned it into incomplete after the first concept NACKs), and for SendMessages, which is not consensus code and it's now outside main/validation. Anyway, any of that is certainly for 0.15.

jonasschnelli commented:
utACK https://github.com/bitcoin/bitcoin/commit/376b3c2c6e329357e4793c1d1b90d1dc0f30fed0
